### PR TITLE
feature/add-ansible-wrapper

### PR DESCRIPTION
**Pull request by ns-jmoneil**
@adnelson @mwilson-ns 

I've added some stuff so that lsi can dynamically create ansible inventories based on filters and run ad hoc commands (modules) or playbooks against those hosts.

I'm sorry ahead of time for the hackishness of this. I'm sure you'll have tons of comments

### CODE REVIEW

**Comment by ns-jmoneil** [comment]
sample module:
python ./lsi.py (search-filter) --latest --ansible --module command --params 'uname -r'

sample playbook:
python ./lsi.py (search-filter) --latest --ansible --playbook playbook.yaml

**Comment by ns-jmoneil** [comment]
@sfoulkes @jdrake

File: src/lsi/utils/ansiblelsi.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
+# Copyright (c) 2015, Narrative Science
```
Comment (sfoulkes) [inline src/lsi/utils/ansiblelsi.py]: Can you name this module ansible_lsi

File: src/lsi/utils/ansiblelsi.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+import ansible.playbook
```
Comment (sfoulkes) [inline src/lsi/utils/ansiblelsi.py]: You'll need to add ansible to your requirements.txt

File: requirements.txt [hunk 3]
```diff
@@ -1,2 +1,3 @@
 boto>=2.32.1
 colored>=1.1.5
+ansible>=1.9.3
```
Comment (jdrake) [inline requirements.txt]: sort the lines plz

File: src/lsi/lsi.py [hunk 4]
```diff
@@ -612,6 +621,31 @@ def main(progname=sys.argv[0]):
         attribs = HostEntry.list_attributes()
         print 'The following attributes are available: {}'\
                     .format(', '.join(attribs))
+    elif args.ansible is not None:
+        latest = True
+        d = {}
```
Comment (jdrake) [inline src/lsi/lsi.py]: use a more descriptive var name

File: src/lsi/lsi.py [hunk 5]
```diff
@@ -612,6 +621,31 @@ def main(progname=sys.argv[0]):
...
+        if profile.filters > 1:
+            query_group = '-'.join(profile.filters)
+            d[query_group] = []
+        elif profile.filters == 1:
+            query_group = profile.filters
+            d[query_group] = []
+        else:
+            query_group = ''
+            d[query_group] = []
```
Comment (jdrake) [inline src/lsi/lsi.py]: this line happens in every case so do it once after the if/else block

File: src/lsi/lsi.py [hunk 6]
```diff
@@ -612,6 +621,31 @@ def main(progname=sys.argv[0]):
...
+        for i in HostEntry.ansible_render_entries(entries, additional_columns=args.show,
```
Comment (jdrake) [inline src/lsi/lsi.py:634]: You only need to assign this once (omit the `d[query_group] = []`):

``` python
d[query_group] = HostEntry.ansible_render_entries(
    entries, additional_columns=args.show, only_show=args.only)
```

File: src/lsi/lsi.py [hunk 7]
```diff
@@ -612,6 +621,31 @@ def main(progname=sys.argv[0]):
...
+                                                only_show=args.only):
+            d[query_group].append(i)
+        if args.playbook is not None:
+            inv_file = build_inventory(d,query_group)
```
Comment (jdrake) [inline src/lsi/lsi.py]: add a space after function arguments

File: src/lsi/lsi.py [hunk 8]
```diff
@@ -612,6 +621,31 @@ def main(progname=sys.argv[0]):
...
+            result = run_playbook(inv_file,args.playbook,query_group)
+        elif args.module is not None:
+            inv_file = build_inventory(d,query_group)
+            result = run_module(inv_file, query_group, args.module,
```
Comment (jdrake) [inline src/lsi/lsi.py]: pep8 formatting here would be:

``` python
result = run_module(
    inv_file, query_group, args.module, module_params=args.params)
```

File: src/lsi/utils/ansible_lsi.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+import ansible.playbook
+import ansible.runner
+import ansible.inventory
+from ansible import callbacks
+from ansible import utils
+import json
```
Comment (jdrake) [inline src/lsi/utils/ansible_lsi.py]: sort imports in blocks: 
1. stdlib (json, os)
2. external dependencies (ansible)
3. local modules (N/A here)

File: src/lsi/utils/ansible_lsi.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    :param inventory_dict: The dictionary to create inventory
+    :type inventory_dict: ``dict``
+    :param query_group: Subset group created from lsi filters
+    :type query_group: ``str``
+    
+    :return: Inventory file name
+    :rtype: ``str``
+    '''
+    _INVENTORY_FILE = os.path.expanduser('~/.lsi-ansible-inventory')
+    inventory_file = _INVENTORY_FILE
```
Comment (sfoulkes) [inline src/lsi/utils/ansible_lsi.py]: You don't need the all caps variable.  All caps implies it's a module level constant.

File: src/lsi/utils/ansible_lsi.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    with open(inventory_file, 'w+') as f:
+        for key in inventory_dict.iterkeys():
+            f.write('['+key+']\n')
+        for val in inventory_dict.itervalues():
+            f.write('\n'.join(val))
+            #print('\n'.join(inventory_dict[query_group]))
```
Comment (sfoulkes) [inline src/lsi/utils/ansible_lsi.py]: rm

File: src/lsi/utils/ansible_lsi.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    :type query_group: ``str``
+    
+    :return: Inventory file name
+    :rtype: ``str``
+    '''
+    _INVENTORY_FILE = os.path.expanduser('~/.lsi-ansible-inventory')
+    inventory_file = _INVENTORY_FILE
+    with open(inventory_file, 'w+') as f:
+        for key in inventory_dict.iterkeys():
+            f.write('['+key+']\n')
```
Comment (sfoulkes) [inline src/lsi/utils/ansible_lsi.py]: Use a format string here:

```
f.write('[{0}]\n'.format(key)
```

Comment (sfoulkes) [inline src/lsi/utils/ansible_lsi.py]: Also don't include a syntax error like my example did.

File: src/lsi/utils/ansible_lsi.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    :type subset: ``str``
+    
+    :return: Output from ansible playbook runner
+    :rtype: ``str``
+    '''
+    inv = ansible.inventory.Inventory(inventory_file)
+    inv.subset(subset)
+    inv.set_playbook_basedir(os.path.dirname(playbook_file))
+    hosts = inv.list_hosts(subset)
+    failed_hosts = list()
```
Comment (sfoulkes) [inline src/lsi/utils/ansible_lsi.py]: Are hosts or failed_hosts used anywhere?

Comment (ns-jmoneil) [inline src/lsi/utils/ansible_lsi.py]: They were going to be, but no they aren't

File: README.md [hunk 14]
```diff
@@ -210,6 +210,77 @@ filters=stg,database,data
 
 Then the `lsi -p stg-database` command will be equivalent to the above.
 
+#### Printing JSON output
+
+You can print JSON output with `lsi` by passing through the `--ansible` option with no other ansible specific options. The ansible options include `--playbook` and `--module`. `--module` can take the optional `--params`.
```
Comment (adnelson) [inline README.md:215]: What does this mean, to print JSON output? And why would this be under the `--ansible` option?

Comment (adnelson) [inline README.md:215]: If you're describing the `--ansible` options you should probably put those in their own section

File: README.md [hunk 15]
```diff
@@ -210,6 +210,77 @@ filters=stg,database,data
...
+
+`lsi` will format the search filters for use as the key in the dictionary. If multiple filters are provided, they are combined into a single string.
```
Comment (adnelson) [inline README.md:217]: Can you give an example of this? I'm not sure what this means

File: README.md [hunk 16]
```diff
@@ -210,6 +210,77 @@ filters=stg,database,data
...
+        }
+    },
+    "dark": {}
+}
+```
+
+`--params` should be used with modules that require parameters
+
+```
+> lsi database --ansible --module command --params 'uptime'
```
Comment (adnelson) [inline README.md:244]: What about key-value params?

File: src/lsi/lsi.py [hunk 17]
```diff
@@ -529,6 +530,14 @@ def _get_args():
                         help='Query AWS for latest instances')
     parser.add_argument('--version', action='store_true', default=False,
                         help='Print version and exit')
+    parser.add_argument('--ansible', action='store_true', default=None,
```
Comment (adnelson) [inline src/lsi/lsi.py:533]: It would probably be better to add a subcommand for this.

File: src/lsi/lsi.py [hunk 18]
```diff
@@ -612,6 +621,29 @@ def main(progname=sys.argv[0]):
...
+        latest = True
+        inventory_list = {}
+        if profile.filters > 1:
+            query_group = '-'.join(profile.filters)
+        elif profile.filters == 1:
+            query_group = profile.filters
+        else:
+            query_group = 'all'
+        inventory_list[query_group] = []
+        for i in HostEntry.ansible_render_entries(entries, additional_columns=args.show,
```
Comment (adnelson) [inline src/lsi/lsi.py:634]: line length

File: src/lsi/lsi.py [hunk 19]
```diff
@@ -612,6 +621,29 @@ def main(progname=sys.argv[0]):
         attribs = HostEntry.list_attributes()
         print 'The following attributes are available: {}'\
                     .format(', '.join(attribs))
+    elif args.ansible is not None:
+        latest = True
+        inventory_list = {}
```
Comment (adnelson) [inline src/lsi/lsi.py:626]: Why is this being used? It looks like there's only one `query_group` being made anyway

Also, it's not a list

File: src/lsi/lsi.py [hunk 20]
```diff
@@ -612,6 +621,29 @@ def main(progname=sys.argv[0]):
...
+        if profile.filters > 1:
+            query_group = '-'.join(profile.filters)
+        elif profile.filters == 1:
+            query_group = profile.filters
```
Comment (adnelson) [inline src/lsi/lsi.py:630]: probably you'd mean `query_group = profile.filters[0]`

File: src/lsi/lsi.py [hunk 21]
```diff
@@ -612,6 +621,29 @@ def main(progname=sys.argv[0]):
...
+        else:
+            query_group = 'all'
+        inventory_list[query_group] = []
+        for i in HostEntry.ansible_render_entries(entries, additional_columns=args.show,
+                                                only_show=args.only):
+            inventory_list[query_group].append(i)
+        if args.playbook is not None:
```
Comment (adnelson) [inline src/lsi/lsi.py:637]: It should probably be an error if these are set and `--ansible` isn't.

File: src/lsi/utils/ansible_lsi.py [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+    inventory_file = os.path.expanduser('~/.lsi-ansible-inventory')
+    with open(inventory_file, 'w+') as f:
+        for key in inventory_dict.iterkeys():
+            f.write('[{}]\n'.format(key))
+        for val in inventory_dict.itervalues():
+            f.write('\n'.join(val))
+        f.write('\n')
+    return inventory_file
+
+def run_playbook(inventory_file,playbook_file,subset):
```
Comment (adnelson) [inline src/lsi/utils/ansible_lsi.py:61]: space after commas

File: src/lsi/utils/ansible_lsi.py [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+    :param inventory_dict: The dictionary to create inventory
+    :type inventory_dict: ``dict``
+    :param query_group: Subset group created from lsi filters
+    :type query_group: ``str``
+    
+    :return: Inventory file name
+    :rtype: ``str``
+    '''
+    inventory_file = os.path.expanduser('~/.lsi-ansible-inventory')
+    with open(inventory_file, 'w+') as f:
```
Comment (adnelson) [inline src/lsi/utils/ansible_lsi.py:53]: Why is this appending each time?

File: src/lsi/lsi.py (hunk 17 above)
Comment (mwilson-ns) [inline src/lsi/lsi.py:533]: :+1:

**Pull request closed without merging**